iree/compiler/Dialect/HAL/Target/LLVM/StaticLibraryGenerator.cpp

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <algorithm>
+#include <cstring>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+namespace {
+
+void generateIntroComment(llvm::raw_ostream &os,
+                          const std::string &library_name,
+                          const std::string &query_function_name) {
+  os << "// === [" << library_name << "] static library ===\n";
```

Comment:
I found it hard to get this to look right in code, since I have to space multi lines exactly, ignoring indentation. The following comments about chaining streams definitely cleaned this up.